os/lib/newlib/syscalls.c

```diff
@@ -76,7 +76,7 @@ _sbrk(int incr)
   static uint8_t *heap_end = &_heap;
   uint8_t *prev_heap_end = heap_end;
 
-  if(heap_end + incr > &_eheap) {
+  if((uintptr_t)heap_end + incr > (uintptr_t)&_eheap) {
```

Comment:
Can you explain the need to cast pointer values to `uintptr_t`, here and in other places?

Reply:
Cppcheck v2.3 says:
```
    "severity": "critical",
    "check_name": "comparePointers",
    "description": "[cppcheck] Comparing pointers that point to different objects (CWE-570)",
```

With the point being that doing relational comparison on pointers to different memory spaces is undefined behavior - while comparing `uintptr_t`/`intptr_t` (which is intended to be used for integer presentation of pointers) is not. However, I see two problems now: 1) After a bit more reading it seems we must actually first cast to `void*` because that is the only type the standard guarantees may be casted to `uintptr_t`/`intptr_t`, and 2) Sky and z1 now complains. Not sure why, but these types are optional...

Reply:
The failure on Sky/Z1 was because they were configured to compile for an older standard than C99, that has been fixed.

I would have expected the pointers to be char * rather than uint8_t * in this code, and upstream newlib seems to both do that and have updated the code to also remove the non-standard caddr_t type: 

https://sourceware.org/git/?p=newlib-cygwin.git;a=blob_plain;f=libgloss/arm/syscalls.c;hb=refs/heads/master

Does cppcheck still warn for the updated upstream function?

Reply:
Thank you for the input @pjonsson . I will re-visit this in the not too distance future.